### PR TITLE
T1 fuzzing fixes

### PR DESCRIPTION
**Pull request by andrewkozlik**
This PR fixes some issues discovered by @invd during fuzzing. There are no security implications AFAICS. Also no UX implications, because the issues manifest only on invalid inputs.
- Potential division by zero in the progress dialog: Trezor One is configured to return 0 and continue execution, so the progress will just show up as 0.
- Potential integer overflows:
  - The depth parameter in BIP32 XPUB: The parameter is merely informational, it does not play any security role. BIP-32 doesn't define how to handle depth > 2^32, so technically it's not even a bug, but an error should be returned in this case, because there is obviously something wrong with the input.
  - `quantity` in `nem_canonicalizeMosaics()`: A bug that doesn't appear to be exploitable, namely because the canonization happens prior to user confirmation. Trezor should emit an error, because it indicates invalid input.
  - Fee rate in `formatComputedFeeRate()`: For the overflow to happen, the fee would have to be more than 1844 million BTC and the user would actually have to be transferring more than 1844 million BTC, but there are less than 21 million BTC in existence. Even if some Bitcoin-like altcoin had more units in existence, there is the `layoutFeeOverThreshold()` warning and the user would still see the absolute fee, which would actually push the fee rate out of the screen.
- Zero-length `pubkeys` VLA in `cryptoMultisigPubkeyIndex()`: The array is never accessed in this case nor is its address taken.

### CODE REVIEW

File: crypto/bip32.c [hunk 1]
```diff
@@ -256,6 +256,10 @@ int hdnode_private_ckd_bip32(HDNode *inout, uint32_t i) {
     memcpy(inout->private_key, I, 32);
   }
 
+  if (inout->depth == UINT32_MAX) {
```
Comment (romanz) [inline crypto/bip32.c]: Can we exit this method earlier in case `inout->depth == UINT32_MAX`?
(also below)

File: legacy/firmware/layout2.c [hunk 2]
```diff
@@ -581,6 +581,10 @@ static bool formatComputedFeeRate(uint64_t fee, uint64_t tx_weight,
   // Compute fee rate and modify it in place for the bn_format_amount()
   // function. We multiply by 100, because we want bn_format_amount() to display
   // two decimal digits.
+  if (100 * fee < fee) {
```
Comment (romanz) [inline legacy/firmware/layout2.c]: Can we add device tests for the new overflow handling code?

File: legacy/firmware/trezor.c [hunk 3]
```diff
@@ -117,11 +118,20 @@ void check_lock_screen(void) {
   }
 }
 
+void trezor_set_busy(uint32_t length_ms) {
+  system_millis_busy_start = timer_ms();
+  system_millis_busy_length = length_ms;
+}
+
+bool trezor_is_busy(void) {
+  return (uint32_t)(timer_ms() - system_millis_busy_start) <
```
Comment (romanz) [inline legacy/firmware/trezor.c]: IIUC, `timer_ms()`'s return value and `system_millis_busy_start` are both `uint32_t` - so I am not sure if we need to cast the result to `uint32_t`.

File: legacy/firmware/signing.c [hunk 4]
```diff
@@ -392,17 +392,32 @@ static void report_progress(bool force) {
   if (progress_midpoint != 0) {
     if (progress_step < progress_midpoint) {
       // Checking previous transactions.
-      progress =
-          (500 * progress_step + 500 * progress_substep / progress_substeps) /
-          progress_midpoint;
+      if (progress_substeps != 0) {
```
Comment (romanz) [inline legacy/firmware/signing.c]: Maybe we can rewrite it using the following "interpolation" function?
```c
uint32_t interpolate(uint32_t value, uint32_t num, uint32_t den) {
    if (den == 0) {
        return value;
    }
    return (value * num) / den;
}

static void report_progress(bool force) {
  static uint32_t update_ctr = 0;
  if (!force && update_ctr < progress_update) {
    update_ctr++;
    return;
  }

  uint32_t progress = 0;
  if (progress_midpoint != 0) {
    if (progress_step < progress_midpoint) {
      // Checking previous transactions.
      progress = (500 * progress_step + interpolate(500, progress_substep, progress_substeps)) / progress_midpoint;
    } else {
      // Signing transaction after checking. No substeps.
      progress = 500 + interpolate(500, progress_step - progress_midpoint, progress_steps - progress_midpoint);
    }
  } else {
    // Loading transaction or signing transaction without checking. No substeps.
    progress = interpolate(1000, progress_step, progress_steps);
  }

  layoutProgress(progress_label, progress);
  update_ctr = 0;
}
```

File: legacy/firmware/trezor.c (hunk 3 above)
Comment (andrewkozlik) [reply legacy/firmware/trezor.c]: https://github.com/trezor/trezor-firmware/pull/6199/commits/237dae4ef67f0aca1b0deb477ed663c3c6b21437

File: legacy/firmware/signing.c (hunk 4 above)
Comment (andrewkozlik) [reply legacy/firmware/signing.c]: Good idea https://github.com/trezor/trezor-firmware/pull/6199/commits/7d6ba22e9f0e461576a6907abb51f8cbe2c79c54

File: crypto/bip32.c (hunk 1 above)
Comment (andrewkozlik) [reply crypto/bip32.c]: https://github.com/trezor/trezor-firmware/pull/6199/commits/483c4883a43ac60df029062e66a7152c507ebff3

File: legacy/firmware/signing.c [hunk 5]
```diff
@@ -381,6 +381,13 @@ static bool is_external_input(uint32_t i) {
   return external_inputs[i / 32] & (1 << (i % 32));
 }
 
+uint32_t interpolate(uint32_t value, uint32_t num, uint32_t den) {
```
Comment (romanz) [inline legacy/firmware/signing.c]: nit:
```suggestion
static uint32_t interpolate(uint32_t value, uint32_t num, uint32_t den) {
```

Comment (andrewkozlik) [reply legacy/firmware/signing.c]: https://github.com/trezor/trezor-firmware/pull/6199/commits/5c17aeee15bc520a4b37b28acfc49514dc38b8d6

File: legacy/firmware/layout2.c (hunk 2 above)
Comment (andrewkozlik) [reply legacy/firmware/layout2.c]: https://github.com/trezor/trezor-firmware/pull/6199/commits/86ca83223cb4c9bcc6d02b01035aa6e78828899b

File: legacy/firmware/layout2.c [hunk 6]
```diff
@@ -581,11 +581,18 @@ static bool formatComputedFeeRate(uint64_t fee, uint64_t tx_weight,
...
   // function. We multiply by 100, because we want bn_format_amount() to display
   // two decimal digits.
-  uint64_t fee_rate_multiplied = div_round(100 * fee, tx_size);
-
-  size_t length =
-      bn_format_amount(fee_rate_multiplied, parentheses ? "(" : NULL,
-                       segwit ? " sat/vB" : " sat/B", 2, output, output_length);
+  int decimals = 2;
+  int multiplier = 100;
+  if (multiplier * fee < fee) {
```
Comment (romanz) [inline legacy/firmware/layout2.c]: Won't it still cause an overflow to be detected during fuzzing?

Also, not sure if it covers all overflow cases, e.g.:
```python
>>> x = 1 << 61
>>> y = (100*x) % (1 << 64)
>>> (x/1e15, y/1e15, y < x)
(2305.843009213694, 9223.372036854777, False)
```

Maybe it's simpler to change the `if` to use 0 decimals if the fee is above some large constant?
```suggestion
  if (fee >= (UINT64_MAX / 100)) {
```

File: tests/device_tests/bitcoin/test_signtx.py [hunk 7]
```diff
@@ -732,6 +732,46 @@ def test_fee_high_hardfail(session: Session):
...
+                (is_core(session), messages.ButtonRequest(code=B.ConfirmOutput)),
+                messages.ButtonRequest(code=B.FeeOverThreshold),
+                messages.ButtonRequest(code=B.SignTx),
+                request_input(0),
+                request_output(0),
+                request_input(0),
+                request_finished(),
+            ]
+        )
+        btc.sign_tx(
```
Comment (romanz) [inline tests/device_tests/bitcoin/test_signtx.py:767]: IIUC, it fails on Core devices with `DataError: The fee is unexpectedly large`:
https://github.com/trezor/trezor-firmware/actions/runs/19934893920/job/57157426225?pr=6199

**Review by romanz — APPROVED** [review]
LGTM, thanks!

File: legacy/firmware/layout2.c (hunk 6 above)
Comment (andrewkozlik) [reply legacy/firmware/layout2.c]: > Won't it still cause an overflow to be detected during fuzzing?

Not sure. We have been using this overflow detection mechanism in other places, such as `add_amount()` in `signing.c`. 

Good catch that it doesn't work in all cases! This logic works for addition, not multiplication. https://github.com/trezor/trezor-firmware/pull/6199/commits/28e25510b6da409cac2fcec6ca69ecb8a7954cca

File: tests/device_tests/bitcoin/test_signtx.py (hunk 7 above)
Comment (andrewkozlik) [reply tests/device_tests/bitcoin/test_signtx.py:767]: Restricted the test to legacy. (We already have `test_fee_high_hardfail` for core and it doesn't make sense to mix them IMO.) https://github.com/trezor/trezor-firmware/pull/6199/commits/28e25510b6da409cac2fcec6ca69ecb8a7954cca
